plutus-wallet-api/ledger/Ledger/Typed/Constraints.hs

```diff
@@ -0,0 +1,142 @@
+  , TypedOrUntypedTxOuts(..)
+  , toTypedTxConstraints
+  , Ledger.Typed.Constraints.addTypedTxIn
+  , Ledger.Typed.Constraints.addTypedTxOut
+  , toUntypedLedgerConstraints
+  ) where
+
+import           Data.Either                 (partitionEithers)
+import           Data.Kind
+import           Data.Set                    (Set)
+import qualified Data.Set                    as Set
+
+import           Ledger.Constraints
+import           Ledger.Constraints.OnChain  (OnChainUtxo (..))
+import           Ledger.Constraints.UtxoLike (UtxoLike (..))
+import           Ledger.Scripts              (DataValue (..))
+import           Ledger.Tx                   (TxIn, TxOut)
+import           Ledger.Typed.Scripts        (ScriptInstance, ScriptType (..))
+import           Ledger.Typed.Tx
+import           Ledger.Typed.TypeUtils
+
+import           Language.PlutusTx
+
+data TypedOrPubKeyTxIns (ins :: [Type]) =
+  TypedOrPubKeyTxIns
+    { topTypedTxIns  :: HListF TypedScriptTxIn ins
+    , topPubKeyTxIns :: [PubKeyTxIn]
+    }
+
+untypedInputs :: forall ins. TypedOrPubKeyTxIns ins -> Set TxIn
+untypedInputs TypedOrPubKeyTxIns{topTypedTxIns, topPubKeyTxIns} =
+  Set.fromList (hfOut tyTxInTxIn topTypedTxIns)
+  <> Set.fromList (unPubKeyTxIn <$> topPubKeyTxIns)
+
+addTypedTxIn'
+  :: forall (ins :: [Type]) inn
+  .  TypedScriptTxIn inn
+  -> TypedOrPubKeyTxIns ins
+  -> TypedOrPubKeyTxIns (inn ': ins)
+addTypedTxIn' txi t =
+  t { topTypedTxIns = HConsF txi (topTypedTxIns t) }
+
+data TypedOrUntypedTxOuts (outs :: [Type]) =
+  TypedOrUntypedTxOuts
+    { totTypedTxOuts :: HListF TypedScriptTxOut outs
+    , totOtherOuts   :: [TxOut]
+    }
+
+addTypedTxOut'
+  :: forall (outs :: [Type]) out
+  .  TypedScriptTxOut out
+  -> TypedOrUntypedTxOuts outs
+  -> TypedOrUntypedTxOuts (out ': outs)
+addTypedTxOut' txo t =
+  t { totTypedTxOuts = HConsF txo (totTypedTxOuts t) }
+
+untypedOutputs :: forall outs. TypedOrUntypedTxOuts outs -> [TxOut]
+untypedOutputs TypedOrUntypedTxOuts{totTypedTxOuts, totOtherOuts} =
+  hfOut tyTxOutTxOut totTypedTxOuts ++ totOtherOuts
+
+untypedDataValues :: forall outs. TypedOrUntypedTxOuts outs -> [DataValue]
+untypedDataValues TypedOrUntypedTxOuts{totTypedTxOuts} = hfOut dv totTypedTxOuts where
+  dv TypedScriptTxOut{tyTxOutData} = DataValue (toData tyTxOutData)
+
+type TypedTxConstraints (ins :: [Type]) (outs :: [Type]) =
+  TxConstraints (TypedOrPubKeyTxIns ins) (TypedOrUntypedTxOuts outs)
+
+emptyTypedTxConstraints :: TypedTxConstraints '[] '[]
+emptyTypedTxConstraints =
+  (mempty @(TxConstraints () ())) { tcInputs = TypedOrPubKeyTxIns HNilF [], tcOutputs = TypedOrUntypedTxOuts HNilF [] }
+
+-- | Turn an on-chain 'PendingTxConstraints' value into an (off-chain)
+--   'TypedTxConstraints '[]' value with either zero or one
+--   continuing outputs. Returns 'Nothing' if the self-payment is overspecified
+--   (more than one output to the contract's own address). All outputs other
+--   than the self-payment remain untyped.
```

Comment:
Ah, now I see why we need this. We don't type the script outputs, so they have to go into a generic bag. I'm not wild about that, the whole point of the typed constraints is that the script outputs are accounted for explicitly!

Reply:
The problem is what happens when we have a constraint about paying to a script address other than our own address. We'd need something like `TypedTxConstraints` in on-chain code.